src/gsCore/gsMemory.h

```diff
@@ -195,6 +195,10 @@ inline shared_ptr<T> make_shared_not_owned(const T *x)
 template <typename T>
 inline unique_ptr<T> make_unique(T * x) { return unique_ptr<T>(x); }
 
+/// Takes a T*, cast it to C* and wraps it in an unique_ptr.
+template <class C, typename T>
+inline unique_ptr<C> make_unique(T * x) { return unique_ptr<C>(dynamic_cast<C*>(x)); }
```

Comment:
what is the use of this ?
better commit it with some usage.
Also, dynamic_cast might fail and then we get a nullptr..

Reply:
i am also not happy with that; we should try not to hide *explicit* casts.

If c++ allows an implicit cast, everything works fine, so we can already write

make_unique<Base>( new Derived() );

Reply:
Need a way to get a raw pointer from one type direct to anthers uPtr.

Happens in segment2d_domain_example.cpp#36:
gsBSpline<real_t>  b =*safe(dynamic_cast<gsBSpline<real_t> * > ( P->outer().singleCurve() ));

after refactor it would be something like:
gsBSpline<real_t>::uPtr b = memory::convert_ptr<gsBSpline<real_t> >( memory::make_unique(P->outer().singleCurve()) );

with this:
gsBSpline<real_t>::uPtr b = memory::make_unique<gsBSpline<real_t> >(P->outer().singleCurve());

Reply:
we do not yet have convert_ptr yet?

c++11 has some

std::static_pointer_cast
std::dynamic_pointer_cast
std::const_pointer_cast

i would follow them.

and if P->outer().singleCurve() would return a uPtr - which it should -, we could have

gsBSpline<real_t>::uPtr b = memory::dynamic_pointer_cast<gsBSpline<real_t> >(P->outer().singleCurve());

Reply:
gsCurve<T> * gsCurveLoop<T>::singleCurve() const

would prefer a uPtr instead also.. should I do?

Reply:
i would say yes.

Reply:
> If c++ allows an implicit cast, everything works fine, so we can already write
> make_unique( new Derived() );
Isn't that nearly what std::make_unique (C++14) does?

Reply:
actually std::make_unique and std::make_shared are *completely* different to our memory::make_unique and memory::make_shared.

Ours take raw pointers and wrap them into the pointer type, std's functions also construct the object and take the constructor's arguments as their arguments.

see also http://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique